src/query/functions/temporal/rate.go

```diff
@@ -36,42 +37,180 @@ const (
-	if optype == IRateType || optype == IDeltaType {
-		return newBaseOp(args, optype, newRateNode, nil)
-	}
+	// RateType calculates the per-second average rate of increase of the time series.
+	RateType = "rate"
 
-	return nil, fmt.Errorf("unknown rate type: %s", optype)
-}
+	// DeltaType calculates the difference between the first and last value of each time series.
+	DeltaType = "delta"
+
+	// IncreaseType calculates the increase in the time series.
+	IncreaseType = "increase"
+)
 
-func newRateNode(op baseOp, controller *transform.Controller, opts transform.Options) Processor {
-	isRate := op.operatorType == IRateType
+type rateProcessor struct {
+	isRate, isCounter bool
+	rateFn            rateFn
+}
 
+func (r rateProcessor) Init(op baseOp, controller *transform.Controller, opts transform.Options) Processor {
 	return &rateNode{
 		op:         op,
 		controller: controller,
 		timeSpec:   opts.TimeSpec,
-		isRate:     isRate,
+		isRate:     r.isRate,
+		isCounter:  r.isCounter,
+		rateFn:     r.rateFn,
 	}
 }
 
+// NewRateOp creates a new base temporal transform for rate functions
+func NewRateOp(args []interface{}, optype string) (transform.Params, error) {
+	var (
+		isRate, isCounter bool
+		rateFn            = standardRateFunc
+	)
+
+	switch optype {
+	case IRateType:
+		isRate = true
+		rateFn = irateFunc
+	case IDeltaType:
+		rateFn = irateFunc
+	case RateType:
+		isRate = true
+		isCounter = true
+	case IncreaseType:
+		isCounter = true
+	case DeltaType:
+	default:
+		return nil, fmt.Errorf("unknown rate type: %s", optype)
+	}
+
+	r := rateProcessor{
+		isRate:    isRate,
+		isCounter: isCounter,
+		rateFn:    rateFn,
+	}
+
+	return newBaseOp(args, optype, r)
+}
+
+type rateFn func(ts.Datapoints, bool, bool, transform.TimeSpec, time.Duration) float64
+
 type rateNode struct {
-	op         baseOp
-	controller *transform.Controller
-	timeSpec   transform.TimeSpec
-	isRate     bool
+	op                baseOp
+	controller        *transform.Controller
+	timeSpec          transform.TimeSpec
+	isRate, isCounter bool
+	rateFn            rateFn
 }
 
 func (r *rateNode) Process(datapoints ts.Datapoints) float64 {
-	valuesLen := len(datapoints)
-	if valuesLen < 2 {
+	return r.rateFn(datapoints, r.isRate, r.isCounter, r.timeSpec, r.op.duration)
+}
+
+func standardRateFunc(
+	datapoints ts.Datapoints,
+	isRate, isCounter bool,
+	timeSpec transform.TimeSpec,
+	timeWindow time.Duration) float64 {
+	if len(datapoints) < 2 {
+		return math.NaN()
+	}
+
+	var (
+		counterCorrection   float64
+		firstVal, lastValue float64
+		firstIdx, lastIdx   int
+		firstTS, lastTS     time.Time
+		foundFirst          bool
+	)
+
+	for i, dp := range datapoints {
+		if math.IsNaN(dp.Value) {
+			continue
+		}
+
+		if !foundFirst {
+			firstVal = dp.Value
+			firstTS = dp.Timestamp
+			firstIdx = i
+			foundFirst = true
+		}
+
+		if isCounter && dp.Value < lastValue {
+			counterCorrection += lastValue
```

Comment:
This will always hit when you're setting first. Is this the correct behaviour?

Reply:
This is what prom does so going to leave it.